doc/README.md

```diff
@@ -0,0 +1,421 @@
+by using HKDF ([RFC5869](https://tools.ietf.org/html/rfc5869)).
+These second 256 bits must not be used for any other purpose in the future.
+We use them for a main key to avoid users having to handle yet another secret.
+
+For deriving keys, we are only using the HKDF's second 'expand' step,
+because the Android Keystore does not give us access
+to the key's byte representation (required for first 'extract' step) after importing it.
+This should be fine as the input key material is already a cryptographically strong key
+(see section 3.3 of RFC 5869 above).
+
+### Key derivation overview
+
+The original entropy comes from a BIP39 seed (12 words = 128 bit size)
+obtained from Java's `SecureRandom`.
+A PBKDF SHA512 based derivation defined in BIP39 turns this into a 512 bit seed key.
+
+The derived seed key (512 bit size) gets split into two parts:
+1. legacy app data encryption key (unused) - 256 bit - first half of seed key
+2. main key - 256 bit - second half of seed key used to derive application specific keys:
+   1. App Backup: HKDF with info "app backup repoId key"
+   2. App Backup: HKDF with info "app backup gear table key"
+   3. App Backup: HKDF with info "app backup stream key"
+   4. File Backup: HKDF with info "stream key"
+   5. File Backup: HKDF with info "Chunk ID calculation"
+
+### Stream Encryption
+
+When a stream is written to backup storage,
+it starts with a header consisting of a single byte indicating the backup format version
+(currently `0x02`) followed by the encrypted payload.
+
+All data written to backup storage will be encrypted with a fresh key
+to prevent issues with nonce/IV re-use of a single key.
+
+We derive a stream key from the main key
+by using HKDF's expand step with the UTF-8 byte representation of "app backup stream key"
+as info input.
+This stream key is then used to derive a new key for each stream.
+
+Instead of encrypting, authenticating and segmenting a cleartext stream ourselves,
+we have chosen to employ the [tink library](https://github.com/tink-crypto/tink-java) for that task.
+Since it does not allow us to work with imported or derived keys,
+we are only using its [AesGcmHkdfStreaming](https://developers.google.com/tink/streaming-aead/aes_gcm_hkdf_streaming)
+to delegate encryption and decryption of byte streams.
+This follows the OAE2 definition as proposed in the paper
+"Online Authenticated-Encryption and its Nonce-Reuse Misuse-Resistance"
+([PDF](https://eprint.iacr.org/2015/189.pdf)).
+
+It adds its own 40 byte header consisting of header length (1 byte), salt and nonce prefix.
+Then it adds one or more segments, each up to 1 MB in size.
+All segments are encrypted with a fresh key that is derived by using HKDF
+on our stream key with another internal random salt (32 bytes) and associated data as info
+([documentation](https://github.com/google/tink/blob/v1.5.0/docs/WIRE-FORMAT.md#streaming-encryption)).
+
+All types of files written to backup storage have the following format:
```

Comment:
[Clarity] If I'm understanding correctly, this is a more-detailed version of a diagram that was shown earlier, so maybe this could read "Below is a more-detailed diagram of the format of all files written to backup storage", or something nicer-sounding to that effect. (Same deal as a few of my other comments.)

Reply:
yeah it seems into the crypto part more, but since it is far away from that other diagram, not sure how to phrase this. Also, a reader my not always go from top to bottom.

Reply:
What about adding a line like this after the previous, zoomed-out diagram:

"The structure of the encrypted tink payload is explored further in Stream Encryption." (where Stream Encryption is a link)?